components/script/dom/htmlimageelement.rs

```diff
@@ -356,20 +390,244 @@ impl HTMLImageElement {
     }
 
     /// <https://html.spec.whatwg.org/multipage/#update-the-source-set>
-    fn update_source_set(&self) -> Vec<DOMString> {
+    fn update_source_set(&self) {
+        // Step 1
+        *self.source_set.borrow_mut() = HTMLImageElement::new_source_set();
+
+        // Step 2
         let elem = self.upcast::<Element>();
-        // TODO: follow the algorithm
-        let src = elem.get_string_attribute(&local_name!("src"));
-        if src.is_empty() {
-            return vec![]
+        let parent = elem.upcast::<Node>().GetParentElement();
+        let is_parent_picture = match parent.clone() {
+            Some(xparent) => xparent.is::<HTMLPictureElement>(),
+            None => false
+        };
+        let nodes = parent.unwrap().upcast::<Node>().ChildNodes();
+        let elements = match is_parent_picture {
+            true =>
+            nodes.iter().filter_map(DomRoot::downcast::<Element>)
+                        .map(|n| DomRoot::from_ref(&*n)).collect(),
+            false => vec![DomRoot::from_ref(&*elem)],
+        };
+
+        // Step 3
+        let _width = match elem.get_attribute(&ns!(), &local_name!("width")) {
+            Some(x) => { match parse_length(&x.value()) {
+                LengthOrPercentageOrAuto::Length(x) => Some(x.to_px() as u32),
+                _ => None
+                }
+            },
+            None => None
+        };
+
+        // Step 4
+        for element in &elements {
+            // Step 4.1
+            if *element == DomRoot::from_ref(&*elem) {
+                let mut source_set = HTMLImageElement::new_source_set();
+                // Step 4.1.1
+                match element.get_attribute(&ns!(), &local_name!("srcset")) {
+                    Some(x) => {
+                        source_set.image_sources = parse_a_srcset_attribute(&x.value());
+                    }
+                    _ => ()
+                }
+
+                // Step 4.1.2
+                match element.get_attribute(&ns!(), &local_name!("sizes")) {
+                    Some(x) => {
+                        source_set.source_size =
+                            parse_a_sizes_attribute(DOMString::from_string(x.value().to_string()));
+                            self.set_value_source_size_list(& mut source_set.source_size);
+                    }
+                    _ => ()
+                }
+
+                // Step 4.1.3
+                let is_str_empty = element.get_string_attribute(&local_name!("src")) ==
+                                         DOMString::from_string("".to_string());
+                let den_img_src_not_found = source_set.image_sources.iter().filter(|imgsource|
+                    match imgsource.descriptor.den {
+                        Some(x) => (x == (1 as f64)),
+                        _ => false
+                    }
+                ).count() == 0;
+                let wid_img_src_not_found = source_set.image_sources.iter().filter(|imgsource|
```

Comment:
We can use `source_set.image_sources.iter().all(|source| source.descriptor.wid.is_none())` and call this `no_width_descriptor`.